[incubator/patroni] PostgreSQL 10 and best practices

hiddeco opened the pull request:
:warning: ~**https://github.com/kubernetes/charts/pull/3953 and https://github.com/kubernetes/charts/pull/3954 should go first**~ :warning:

Applies best practices, defaults to Patroni / Spilo releases based upon PostgreSQL 10 and re-implements admin password functionalities.

hiddeco committed:
Removed obsolete admin password option

There is no Spilo release available based upon Postgresql 9.6 with
a working admin configuration option
(https://github.com/zalando/spilo/pull/189).

Remove the option to not confuse any chart users.

hiddeco committed:
Follow Chart best practices value naming conventions

hiddeco committed:
Fixed imagePullPolicy option

hiddeco committed:
Added recommended 'app' label

https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md

hiddeco committed:
Improved README

hiddeco committed:
Upgrade Spilo image to PostgreSQL 10

hiddeco committed:
Re-implement admin password

hiddeco committed:
Do not depend on default but inherit from values

hiddeco committed:
Be consistent with spacing

hiddeco committed:
Update lock

hiddeco commented:
/assign @unguiculus

jdumars commented:
/ok-to-test

hiddeco committed:
Fixed fullname

hiddeco commented:
/retest

hiddeco committed:
Merge remote-tracking branch 'upstream/master' into patroni-postgresql-10

hiddeco commented:
/retest

unguiculus commented on incubator/patroni/templates/ep-patroni.yaml:
```diff
@@ -3,8 +3,9 @@ kind: Endpoints
 metadata:
   name: {{ template "patroni.fullname" . }}
   labels:
+    app: {{ template "patroni.fullname" . }}
```
`app` labels should be `{{ template "patroni.name" . }}`

unguiculus commented on incubator/patroni/templates/ep-patroni.yaml:
```diff
@@ -3,8 +3,9 @@ kind: Endpoints
     release: {{ .Release.Name | quote }}
     chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
-    component: "{{ .Release.Name }}-{{ .Values.Component }}"
+    component: "{{ .Release.Name }}-{{ .Values.component }}"
```
I'd suggest you remove the release name prefix. A component isn't tied to a release.

unguiculus commented on incubator/patroni/values.yaml:
```diff
@@ -1,55 +1,70 @@
-Resources:
-  Cpu: 100m
-  Memory: 512Mi
+resources:
```
We recommend not to specify defaults for resources.

unguiculus reviewed (changes requested):
Thanks for this. I appreciate your effort. I made a few comments. Also I suggest you look at our review guidelines again for best practices. It'd also be good to add the current `_helpers.tpl` you get with `helm create` using Helm 2.8+.

Have you thought about updating the chart such that it can work without Etcd or Zookeeper? I've played around with this here when support for this was still experimental and not merged yet: https://github.com/unguiculus/charts/tree/feature/patroni/incubator/patroni. It's however a little outdated now.

hiddeco commented:
@unguiculus agree with your comments and will push an update sometime this week.

Would be more than happy to add native Kubernetes support to the chart besides the etcd and ZooKeeper support it already offers. Reason I did not add this yet is that I need to take some time to look in to the functional details and I'd like to keep my PRs in nicely separated chunks of improvements.

I also have to add this has no priority for my employee as the native Kubernetes option has some drawbacks on GKE/GCE due to masters restarting in read-only modus when the Kubernetes API becomes unavailable (https://github.com/zalando/patroni/issues/574#issuecomment-353000821). But as you already paved the road for me in your branch I suspect it will take little of my time.

hiddeco committed:
Update helper functions and finetune labels

hiddeco committed:
Removed component label and refactored values

hiddeco committed:
Mount admin password as env variable

hiddeco committed:
Improve README

hiddeco committed:
Update todo

hiddeco committed:
Document (full)name overrides

hiddeco commented:
@unguiculus I have updated the PR with several improvements:

- Updated the `_helpers.tpl` to match Helm 2.8+
- Updated the labels to match Helm 2.8+
- Dropped the `component` label in favour of the (full)name override it now offers
- Refactored the `values.yaml` one more time:
    - image options are now under the `image` key, also moved the `imagePullPolicy` to `image.pullPolicy`
	- `replicas` -> `replicaCount`
	- Added/refactored `resources`, `nodeSelector`, `tolerations` and `affinity` options
- Added `selector` to the StatefulSet
- Revised the README

hiddeco committed:
Bump Spilo version to 1.4-p6

hiddeco committed:
Use lowercase value names in all if statements

hiddeco committed:
Bump Spilo version to 1.4-p7

hiddeco committed:
Pre-define anti-affinity rules

hiddeco committed:
Merge remote-tracking branch 'upstream/master' into patroni-postgresql-10

hiddeco commented:
/retest

timstoop commented on incubator/patroni/README.md:
```diff
@@ -1,24 +1,25 @@
+This directory contains a Kubernetes chart to deploy a five node [Patroni](https://github.com/zalando/patroni/) cluster using a [Spilo](https://github.com/zalando/spilo) and a StatefulSet.
 
 ## Prerequisites Details
 * Kubernetes 1.5
```
While you're add it, I'd recommend making this `Kubernetes 1.5+`, as I assume it's been tested on newer versions as well.

timstoop commented on incubator/patroni/Chart.yaml:
```diff
@@ -1,7 +1,7 @@
 description: 'Highly available elephant herd: HA PostgreSQL cluster.'
-version: 0.6.3
-appVersion: 1.3-p4
+version: 0.7
```
Please use the full 2-dot notation (0.7.0 in this case).

timstoop commented:
Sorry, didn't mean to do that

timstoop reviewed (changes requested):
One small change request, one small change suggestion.

hiddeco committed:
Use full 2-dot notation and improve README

timstoop commented:
/lgtm

hiddeco commented:
@unguiculus waiting for your approval whenever you find time.

unguiculus commented on incubator/patroni/templates/statefulset-patroni.yaml:
```diff
@@ -3,91 +3,112 @@ kind: StatefulSet
-        {{ end }}
-      {{ end }}
       serviceAccountName: {{ template "patroni.serviceAccountName" . }}
+      affinity:
```
Please allow affinity to be configured the same way as e. g. node selectors or tolerations. This would be more flexible and would allow you to consider e. g. availability zones as well and not just hostnames.

hiddeco replied:
Added the option to set `affinity` rules, if you set these the `antiAffinity` will be ignored.

unguiculus replied:
I don't like this. The `podAntiAffinity` block is rather confusing. You should remove it. You have the full flexibilty to configure anything you need. You still may provide a default as follows using the `tpl` function.

https://github.com/kubernetes/charts/blob/494b80b66acdef9de90bcaf77eb1e5298ff9c0fc/incubator/vault/templates/deployment.yaml#L54
https://github.com/kubernetes/charts/blob/494b80b66acdef9de90bcaf77eb1e5298ff9c0fc/incubator/vault/values.yaml#L40-L49

hiddeco replied:
I wasn't aware of the `tpl` function, updated.

hiddeco committed:
Allow complex affinity rules

hiddeco committed:
Fixed whitelines if/else statements

hiddeco committed:
Update README

hiddeco committed:
Merged affinity and antiAffinity values functionalities

hiddeco committed:
Fix

hiddeco committed:
podAntiAffinity

hiddeco commented:
/retest

hiddeco commented:
@unguiculus what is blocking the merge?

unguiculus commented:
/lgtm
